content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md

```diff
@@ -0,0 +1,249 @@
+---
+title: Manually testing changes to a Paketo Buildpack
+date: "2023-05-23"
+slug: manually-testing-changes-to-paketo-buildpacks
+author: anthonydahanne
+---
+
+*While this article is mainly about how to manually test Java family Paketo buildpacks, still, most of the instructions should work fine with other Paketo buildpacks; and even non Paketo buildpacks!*
+
+Suppose that you started working on a new feature for the [`spring-boot`](https://github.com/paketo-buildpacks/spring-boot/) buildpack...
+
+## Unit tests first 
+
+Of course, during your development, you will make sure that you did not break any tests, regularly running `go test -v ./...` but you would also add new tests covering your new code; the classic combination `go test -v ./... -coverprofile=coverage.out` followed with `go tool cover -html=coverage.out` will allow you to verify you did not forget to test some errors or edge cases.
+
+## Build and try the buildpack out before publishing it
+
+If the buildpack you are working on provides some integration tests (or end to end tests), I suggest you try and add some use cases to those tests.
```

Comment:
Not sure if you want to explain our general testing philosophy with Java buildpacks, I have had questions about it, so it might be worth covering.

1. Test all of the buildpack logic in unit tests. This usually requires tests for individual functions/structs, tests for specific layer providers, and then tests for build and detect.
2. Component buildpacks don't have integration tests. It's difficult to test them in integration because they are usually designed to work together with other buildpacks.
3. Composite buildpacks should have integration tests. This does not need to test every individual feature, because those are covered by 1.) but it should test cross-buildpack compatibility.
4. The Paketo samples offer an additional backstop with additional smoke testing.

While this is the philosophy, we're not quite there. There are a couple of component buildpacks with integration tests (probably should go away) and we've yet to rollout integration tests in the composites (although we approved an RFC to do so).

Reply:
I'm in favor of 3 & 4: I agree integration test at the component bp level is not really interesting.

Integration test should at least cover high level functionality such as "assume gradle spring boot project produces a runnable native image using the `java-native-image` bp composite"

now thinking about it, the samples smoke test (4) do that already...